File: Client/menus/multiplayerSelector/multiplayerSelector.cpp
#include "platform_folders.h"
#include "multiplayerSelector.hpp"
#include "game.hpp"
#include "choiceScreen.hpp"
#include "serverAdder.hpp"
#include "nameChooser.hpp"
#include "readOpa.hpp"
#include "resourcePath.hpp"

#define TOP_HEIGHT (title.getHeight() + 2 * SPACING)
#define BOTTOM_HEIGHT (back_button.h + 2 * SPACING)

void MenuServer::render(int position, int mouse_x, int mouse_y) {
    int render_x = gfx::getWindowWidth() / 2 - 400 + SPACING, render_y = y - position, render_width = 800 - 2 * SPACING, render_height = 116 + 2 * SPACING;
    
    gfx::Color back_color = BLACK;
    back_color.a = 100;
    gfx::RectShape(render_x, render_y, render_width, render_height).render(back_color);
    
    icon.render(1, render_x + SPACING, render_y + SPACING);
    name_texture.render(3, render_x + 2 * SPACING + icon.getTextureWidth(), render_y + SPACING * 1.5);
    
    join_button.x = render_x + 2 * SPACING + icon.getTextureWidth();
    join_button.y = render_y + render_height - join_button.h - SPACING;
    join_button.render(mouse_x, mouse_y);
    
    remove_button.x = render_x + 3 * SPACING + icon.getTextureWidth() + join_button.w;
    remove_button.y = render_y + render_height - join_button.h - SPACING;
    remove_button.render(mouse_x, mouse_y);
}

void MultiplayerSelector::refresh() {
    for(int i = 0; i < servers.size(); i++)
        delete servers[i];
    servers.clear();
    
    scroll_limit = 0;
    
    for(int i = 0; i < server_data.size(); i++) {
        servers.push_back(new MenuServer(ServerData("", "")));
        MenuServer* server = servers.back();
        
        server->y = scroll_limit + TOP_HEIGHT;
        
        server->icon.loadFromSurface(readOpa(resource_path + "world_icon.opa"));
        
        server->data.ip = server_data[i].ip;
        server->data.name = server_data[i].name;
        server->name_texture.loadFromSurface(gfx::textToSurface(server_data[i].name));

        server->join_button.loadFromSurface(readOpa(resource_path + "join_button.opa"));
        server->join_button.setScale(3);
        server->join_button.margin = 5;
        
        server->remove_button.loadFromSurface(readOpa(resource_path + "remove_button.opa"));
        server->remove_button.setScale(3);
        server->remove_button.margin = 5;
        
        scroll_limit += 116 + SPACING * 3;
    }
}

void MultiplayerSelector::init() {
    ConfigFile config(sago::getDataHome() + "/Terralistic/servers.txt");
    config.setDefaultStr("servers", "");
    
    title.setScale(3);
    title.loadFromSurface(gfx::textToSurface("Select a server to join"));
    title.y = SPACING;
    title.orientation = gfx::TOP;

    back_button.setScale(3);
    back_button.loadFromSurface(gfx::textToSurface("Back"));
    back_button.y = -SPACING;
    back_button.orientation = gfx::BOTTOM;

    new_button.setScale(3);
    new_button.loadFromSurface(gfx::textToSurface("Add"));
    new_button.y = -SPACING;
    new_button.orientation = gfx::BOTTOM;
    
    top_rect.orientation = gfx::TOP;
    top_rect.setHeight(TOP_HEIGHT);
    
    bottom_rect.orientation = gfx::BOTTOM;
    bottom_rect.setHeight(BOTTOM_HEIGHT);
    bottom_rect.fill_color.a = TRANSPARENCY / 2;
    bottom_rect.shadow_intensity = SHADOW_INTENSITY;
    bottom_rect.blur_radius = BLUR;
    
    std::string servers_str = config.getStr("servers"), curr_serv_ip, curr_serv_name;
    bool ip_name_switch = false;
    for(int i = 0; i < servers_str.size(); i++) {
        if(!ip_name_switch) {
            if (servers_str[i] == ' ') {
                ip_name_switch = true;
            } else
                curr_serv_name.push_back(servers_str[i]);
        }else{
            if (servers_str[i] == ' ') {
                server_data.emplace_back(curr_serv_name, curr_serv_ip);
                curr_serv_ip.clear();
                curr_serv_name.clear();
                ip_name_switch = false;
            } else
                curr_serv_ip.push_back(servers_str[i]);
        }
    }
    
    refresh();
}

bool MultiplayerSelector::onKeyUp(gfx::Key key) {
    if(key == gfx::Key::MOUSE_LEFT) {
        if(back_button.isHovered(getMouseX(), getMouseY()))
            returnFromScene();
        else if(new_button.isHovered(getMouseX(), getMouseY())) {
            ServerAdder server_adder(menu_back);
            server_adder.run();
            if(!server_adder.server_ip.empty())
                server_data.emplace_back(server_adder.server_name, server_adder.server_ip);
            refresh();
        }
        else
            for(int i = 0; i < servers.size(); i++) {
                if(servers[i]->join_button.isHovered(getMouseX(), getMouseY())) {
                    NameChooser name_chooser(menu_back, settings, servers[i]->data.ip);
                    name_chooser.run();
                } else if(servers[i]->remove_button.isHovered(getMouseX(), getMouseY())) {
                    std::string result;
                    if(getKeyState(gfx::Key::SHIFT))
                        result = "Yes";
                    else {
                        ChoiceScreen choice_screen(menu_back, "Do you want to remove " + servers[i]->data.name + "?", {"Yes", "No"}, &result);
                        choice_screen.run();
                    }

                    if(result == "Yes") {
                        server_data.erase(server_data.begin() + i);
                        refresh();
                    }
                    break;
                }
            }
        return true;
    }
    return false;
}

void MultiplayerSelector::onMouseScroll(int distance) {
    position -= distance * 8;
    if(position < 0)
        position = 0;
    int scroll_limit_ = scroll_limit - gfx::getWindowHeight() + TOP_HEIGHT + BOTTOM_HEIGHT;
    if(position > scroll_limit_)
        position = scroll_limit_ > 0 ? scroll_limit_ : 0;
}

void MultiplayerSelector::render() {
    menu_back->setBackWidth(800);
    menu_back->renderBack();
    
    bool hoverable = getMouseY() > TOP_HEIGHT && getMouseY() < gfx::getWindowHeight() - BOTTOM_HEIGHT;

    for(int i = 0; i < servers.size(); i++) {
        servers[i]->join_button.disabled = !hoverable;
        servers[i]->remove_button.disabled = !hoverable;
    }

    for(int i = 0; i < servers.size(); i++)
        servers[i]->render(position, getMouseX(), getMouseY());

    top_rect.setWidth(menu_back->getBackWidth());
    top_rect_visibility += ((position ? 1.f : 0.f) - top_rect_visibility) / 20;
    if(top_rect_visibility < 0.01f)
        top_rect_visibility = 0;
    if(top_rect_visibility > 0.99f)
        top_rect_visibility = 1;
    top_rect.fill_color.a = top_rect_visibility * TRANSPARENCY / 2;
    top_rect.blur_radius = top_rect_visibility * BLUR;
    top_rect.shadow_intensity = top_rect_visibility * SHADOW_INTENSITY;
    if(top_rect_visibility)
        top_rect.render();
    
    bottom_rect.setWidth(menu_back->getBackWidth());
    int scroll_limit_ = scroll_limit - gfx::getWindowHeight() + TOP_HEIGHT + BOTTOM_HEIGHT;
    if(scroll_limit_ > 0)
        bottom_rect.render();

    title.render();
    back_button.render(getMouseX(), getMouseY());
    
    new_button.x = menu_back->getBackWidth() / 2 - SPACING - new_button.w / 2;
    new_button.render(getMouseX(), getMouseY());
}

void MultiplayerSelector::stop() {
    ConfigFile config(sago::getDataHome() + "/Terralistic/servers.txt");
    std::string servers_str;
    for(int i = 0; i < server_data.size(); i++) {
        servers_str += server_data[i].name;
        servers_str.push_back(' ');
        servers_str += server_data[i].ip;
        servers_str.push_back(' ');
    }
    config.setStr("servers", servers_str);
    for(int i = 0; i < servers.size(); i++)
        delete servers[i];
}

File: Client/menus/multiplayer_selector.rs
use super::world_creation::run_world_creation;
use crate::game::private_world::run_private_world;
use crate::menus::background_rect::BackgroundRect;
use crate::menus::run_choice_menu;

use directories::BaseDirs;
use graphics as gfx;
use graphics::GraphicsContext;
use std::fs;
use std::path::PathBuf;
use serde_derive::{Deserialize, Serialize};


pub const MENU_WIDTH: i32 = 800;

#[derive(Serialize, Deserialize)]
pub struct ServerInfo {
    name: String,
    ip: String,
}

impl ServerInfo {
    pub fn new(name: String, ip: String) -> Self {
        Self { name, ip }
    }
}


/**
World is a struct that contains all information to
render the server in Server selector.
 */
pub struct ServerCard {
    pub server_info: ServerInfo,
    rect: gfx::RenderRect,
    play_button: gfx::Button,
    delete_button: gfx::Button,
    title: gfx::Sprite,
    icon: gfx::Sprite,
}


impl ServerCard {
    pub fn new(graphics: &GraphicsContext, name: String, ip: String) -> Self {

        let mut rect = gfx::RenderRect::new(0.0, 0.0, (MENU_WIDTH - 2 * gfx::SPACING) as f32, 0.0);
        rect.orientation = gfx::TOP;
        rect.fill_color.a = 100;
        rect.smooth_factor = 60.0;

        let mut icon = gfx::Sprite::new();
        icon.texture = gfx::Texture::load_from_surface(&gfx::Surface::deserialize(
            &include_bytes!("../../Build/Resources/world_icon.opa").to_vec(),
        ));
        rect.h = icon.get_height() as f32 + 2.0 * gfx::SPACING as f32;
        icon.x = gfx::SPACING;
        icon.orientation = gfx::LEFT;

        let mut title = gfx::Sprite::new();
        title.texture =
            gfx::Texture::load_from_surface(&graphics.font.create_text_surface(name.clone().as_str()));
        title.x = icon.x + icon.get_width() + gfx::SPACING;
        title.y = gfx::SPACING;
        title.scale = 3.0;

        let mut play_button = gfx::Button::new();
        play_button.texture = gfx::Texture::load_from_surface(&gfx::Surface::deserialize(
            &include_bytes!("../../Build/Resources/play_button.opa").to_vec(),
        ));
        play_button.scale = 3.0;
        play_button.margin = 5;
        play_button.x = icon.x + icon.get_width() + gfx::SPACING;
        play_button.y = -gfx::SPACING;
        play_button.orientation = gfx::BOTTOM_LEFT;

        let mut delete_button = gfx::Button::new();
        delete_button.texture = gfx::Texture::load_from_surface(&gfx::Surface::deserialize(
            &include_bytes!("../../Build/Resources/delete_button.opa").to_vec(),
        ));
        delete_button.scale = 3.0;
        delete_button.margin = 5;
        delete_button.x = play_button.x + play_button.get_width() + gfx::SPACING;
        delete_button.y = -gfx::SPACING;
        delete_button.orientation = gfx::BOTTOM_LEFT;

        Self {
            server_info: ServerInfo::new(name, ip),
            rect,
            play_button,
            delete_button,
            title,
            icon,
        }
    }

    /**
    This function renders the world card on the x and y position.
     */
    pub fn render(
        &mut self, graphics: &mut GraphicsContext, x: i32, y: i32,
        parent_container: Option<&gfx::Container>,
    ) {
        self.rect.x = x as f32;
        self.rect.y = y as f32;
        self.rect.render(graphics, parent_container);

        let rect_container = self.rect.get_container(graphics, parent_container);
        self.icon.render(graphics, Some(&rect_container));
        self.title.render(graphics, Some(&rect_container));
        self.play_button.render(graphics, Some(&rect_container));
        self.delete_button.render(graphics, Some(&rect_container));
    }

    /**
    This function returns height of the world card.
     */
    pub fn get_height(&self) -> i32 {
        self.rect.h as i32
    }

    /**
    This function disables/enables the world card buttons.
     */
    pub fn set_enabled(&mut self, enabled: bool) {
        self.play_button.disabled = !enabled;
        self.delete_button.disabled = !enabled;
    }

    /**
    This function returns the container of the world card.
     */
    pub fn get_container(
        &self, graphics: &GraphicsContext, parent_container: Option<&gfx::Container>,
    ) -> gfx::Container {
        self.rect.get_container(graphics, parent_container)
    }
}

/**
WorldList is a struct that is used to list all worlds in the world folder
and render them in the singleplayer selector menu.
 */
pub struct ServerList {
    pub servers: Vec<ServerCard>,
}

impl ServerList {
    pub fn new(graphics: &GraphicsContext, file_path: PathBuf) -> ServerList {
        let mut world_list = ServerList { servers: Vec::new() };
        world_list.refresh(graphics, file_path);
        world_list
    }

    pub fn refresh(&mut self, graphics: &GraphicsContext, file_path: PathBuf) {
        let temp_servers;

        if !file_path.exists() {
            temp_servers = Ok(Vec::new());
            &std::fs::write(file_path, &bincode::serialize::<&Vec<ServerInfo>>(&temp_servers.as_ref().unwrap()).unwrap()).unwrap();
        } else {
            temp_servers = bincode::deserialize::<Vec<ServerInfo>>(
                &std::fs::read(file_path).unwrap(),
            );
        }

        if temp_servers.is_ok() {
            self.servers.clear();
            for server in temp_servers.unwrap() {
                self.servers.push(ServerCard::new(graphics, server.name, server.ip));
            }
        }
    }
}

pub fn run_multiplayer_selector(
    graphics: &mut GraphicsContext, menu_back: &mut dyn BackgroundRect,
) {

    let base_dirs = BaseDirs::new().unwrap();
    let servers_file = base_dirs.data_dir().join("Terralistic").join("servers.txt");

    let mut server_list = ServerList::new(graphics, servers_file.clone());

    let mut title = gfx::Sprite::new();
    title.scale = 3.0;
    title.texture = gfx::Texture::load_from_surface(
        &graphics
            .font
            .create_text_surface("Select a server to play!"),
    );
    title.y = gfx::SPACING;
    title.orientation = gfx::TOP;

    let mut back_button = gfx::Button::new();
    back_button.scale = 3.0;
    back_button.texture =
        gfx::Texture::load_from_surface(&graphics.font.create_text_surface("Back"));
    back_button.y = -gfx::SPACING;
    back_button.orientation = gfx::BOTTOM;

    let mut new_world_button = gfx::Button::new();
    new_world_button.scale = 3.0;
    new_world_button.texture =
        gfx::Texture::load_from_surface(&graphics.font.create_text_surface("New"));
    new_world_button.y = -gfx::SPACING;
    new_world_button.x = -gfx::SPACING;
    new_world_button.orientation = gfx::BOTTOM_RIGHT;

    let top_height = title.get_height() + 2 * gfx::SPACING;
    let bottom_height = back_button.get_height() + 2 * gfx::SPACING;

    let mut top_rect = gfx::RenderRect::new(0.0, 0.0, 0.0, top_height as f32);
    top_rect.orientation = gfx::TOP;

    let mut bottom_rect = gfx::RenderRect::new(0.0, 0.0, 0.0, bottom_height as f32);
    bottom_rect.fill_color.a = gfx::TRANSPARENCY / 2;
    bottom_rect.shadow_intensity = gfx::SHADOW_INTENSITY;
    bottom_rect.blur_radius = gfx::BLUR;
    bottom_rect.orientation = gfx::BOTTOM;

    let mut top_rect_visibility = 1.0;
    let mut position: f32 = 0.0;

    'render_loop: while graphics.renderer.is_window_open() {
        while let Some(event) = graphics.renderer.get_event() {
            match event {
                gfx::Event::KeyRelease(key, ..) => {
                    match key {
                        gfx::Key::MouseLeft => {
                            if back_button
                                .is_hovered(graphics, Some(menu_back.get_back_rect_container()))
                            {
                                break 'render_loop;
                            }
                            if new_world_button
                                .is_hovered(graphics, Some(menu_back.get_back_rect_container()))
                            {
                                //run_world_creation(graphics, menu_back, &mut world_list.worlds);//TODO: server creation
                                //world_list.refresh(graphics);
                            }

                            let mut needs_refresh = false;
                            for server in &mut server_list.servers {
                                if server.play_button.is_hovered(
                                    graphics,
                                    Some(&server.get_container(
                                        graphics,
                                        Some(menu_back.get_back_rect_container()),
                                    )),
                                ) {
                                    //run_private_world(graphics, menu_back, world.get_file_path());//TODO: join server
                                } else if server.delete_button.is_hovered(
                                    graphics,
                                    Some(&server.get_container(
                                        graphics,
                                        Some(menu_back.get_back_rect_container()),
                                    )),
                                ) && run_choice_menu(format!("The world \"{}\" will be deleted.\nDo you want to proceed?", server.server_info.name), graphics, menu_back, None, None) {
                                    //fs::remove_file(server.get_file_path()).unwrap();//TODO: remove server
                                    needs_refresh = true;
                                }
                            }
                            if needs_refresh {
                                server_list.refresh(graphics, servers_file.clone());
                            }
                        }
                        gfx::Key::Escape => {
                            break 'render_loop;
                        }
                        _ => {}
                    }
                }
                gfx::Event::MouseScroll(delta) => {
                    position += delta * 2.0;
                }
                _ => {}
            }
        }
        menu_back.set_back_rect_width(MENU_WIDTH);

        menu_back.render_back(graphics);

        let hoverable = graphics.renderer.get_mouse_y() as i32 > top_height
            && (graphics.renderer.get_mouse_y() as i32)
            < graphics.renderer.get_window_height() as i32 - bottom_height;

        for server in &mut server_list.servers {
            server.set_enabled(hoverable);
        }

        let mut current_y = gfx::SPACING;
        for i in 0..server_list.servers.len() {
            server_list.servers[i].render(
                graphics,
                0,
                current_y + top_height + position as i32,
                Some(menu_back.get_back_rect_container()),
            );
            current_y += server_list.servers[i].get_height() + gfx::SPACING;
        }

        top_rect.w = menu_back.get_back_rect_width(graphics, None) as f32;
        top_rect_visibility +=
            ((if position < -5.0 { 1.0 } else { 0.0 }) - top_rect_visibility) / 20.0;

        if top_rect_visibility < 0.01 {
            top_rect_visibility = 0.0;
        }

        if top_rect_visibility > 0.99 {
            top_rect_visibility = 1.0;
        }

        top_rect.fill_color.a = (top_rect_visibility * gfx::TRANSPARENCY as f32 / 2.0) as u8;
        top_rect.blur_radius = (top_rect_visibility * gfx::BLUR as f32) as i32;
        top_rect.shadow_intensity = (top_rect_visibility * gfx::SHADOW_INTENSITY as f32) as i32;
        if top_rect_visibility > 0.0 {
            top_rect.render(graphics, Some(menu_back.get_back_rect_container()));
        }

        bottom_rect.w = menu_back.get_back_rect_width(graphics, None) as f32;
        let mut scroll_limit =
            current_y - graphics.renderer.get_window_height() as i32 + top_height + bottom_height;
        if scroll_limit < 0 {
            scroll_limit = 0;
        }

        if scroll_limit > 0 {
            bottom_rect.render(graphics, Some(menu_back.get_back_rect_container()));
        }

        if position > 0.0 {
            position -= position / 20.0;
        }

        if position < -scroll_limit as f32 {
            position -= (position + scroll_limit as f32) / 20.0;
        }

        title.render(graphics, Some(menu_back.get_back_rect_container()));
        back_button.render(graphics, Some(menu_back.get_back_rect_container()));

        new_world_button.render(graphics, Some(menu_back.get_back_rect_container()));

        graphics.renderer.update_window();
    }
}
